wallet/builder.go

```diff
@@ -2,247 +2,364 @@ package wallet
 		if err != nil {
-			return nil, err
+			return InvalidCoinType, nil, err
 		}
-		bitcoinCashPath := path.Join(cfg.RepoPath, "bitcoincash")
-		os.Mkdir(bitcoinCashPath, os.ModePerm)
-		bitcoinCashConfig := &bitcoincash.Config{
-			Mnemonic:             cfg.Mnemonic,
-			Params:               cfg.Params,
-			MaxFee:               uint64(cfg.ConfigFile.BCH.MaxFee),
-			LowFee:               uint64(cfg.ConfigFile.BCH.LowFeeDefault),
-			MediumFee:            uint64(cfg.ConfigFile.BCH.MediumFeeDefault),
-			HighFee:              uint64(cfg.ConfigFile.BCH.HighFeeDefault),
-			FeeAPI:               *feeAPI,
-			RepoPath:             bitcoinCashPath,
-			CreationDate:         cfg.WalletCreationDate,
-			DB:                   CreateWalletDB(cfg.DB, wallet.BitcoinCash),
-			UserAgent:            "OpenBazaar",
-			TrustedPeer:          tp,
-			Proxy:                cfg.Proxy,
-			Logger:               cfg.Logger,
-			DisableExchangeRates: cfg.DisableExchangeRates,
+		return actualCoin, w, nil
+	case wallet.Zcash:
+		if testnet {
+			actualCoin = wallet.TestnetZcash
+		} else {
+			actualCoin = wallet.Zcash
 		}
-		bitcoinCashSPVWallet, err := bitcoincash.NewSPVWallet(bitcoinCashConfig)
+		w, err := zcash.NewZCashWallet(*coinConfig, cfg.Mnemonic, cfg.Params, cfg.Proxy, cache.NewMockCacher(), cfg.DisableExchangeRates)
 		if err != nil {
-			return nil, err
+			return InvalidCoinType, nil, err
 		}
-		if testnet {
-			mw[wallet.TestnetBitcoinCash] = bitcoinCashSPVWallet
+		return actualCoin, w, nil
+	case wallet.Ethereum:
+		actualCoin = wallet.Ethereum
+		w, err := eth.NewEthereumWallet(*coinConfig, cfg.Mnemonic, cfg.Proxy)
+		if err != nil {
+			return InvalidCoinType, nil, err
+		}
+		return actualCoin, w, nil
+	}
+	return InvalidCoinType, nil, fmt.Errorf("unable to create wallet for unknown coin %s", coin)
+}
+
+func createSPVWallet(coin wallet.CoinType, coinConfigOverrides *schema.CoinConfig, cfg *WalletConfig) (wallet.CoinType, wallet.Wallet, error) {
+	var (
+		actualCoin         wallet.CoinType
+		notMainnet         = cfg.Params.Name != chaincfg.MainNetParams.Name
+		usingRegnet        = cfg.Params.Name == chaincfg.RegressionNetParams.Name
+		missingTrustedPeer = coinConfigOverrides.TrustedPeer == ""
+	)
+
+	if usingRegnet && missingTrustedPeer {
+		return InvalidCoinType, nil, ErrTrustedPeerRequired
+	}
+
+	coinConfig, err := prepareSPVCoinConfig(coin, coinConfigOverrides, cfg)
+	if err != nil {
+		return InvalidCoinType, nil, err
+	}
+
+	switch coin {
+	case wallet.Bitcoin:
+		if notMainnet {
+			actualCoin = wallet.TestnetBitcoin
+		} else {
+			actualCoin = wallet.Bitcoin
+		}
+	case wallet.BitcoinCash:
+		if notMainnet {
+			actualCoin = wallet.TestnetBitcoinCash
 		} else {
-			mw[wallet.BitcoinCash] = bitcoinCashSPVWallet
+			actualCoin = wallet.BitcoinCash
 		}
 	}
+	newSPVWallet, err := spvwallet.NewSPVWallet(coinConfig)
```

Comment:
This will need to use `bitcoincash.NewSPVWallet()` if it's a bitcoin cash wallet.